### PR TITLE
CredentialsManager function to clear and revoke the refresh token

### PR DESCRIPTION
**Pull request by stevehobbsdev**
## Changes

This PR adds a new method `revoke` to `CredentialsManager`, which boths clears credentials and revokes the refresh token, if available. It accepts a callback for when the revocation call has finished.

When called, it first checks to make sure that it has:

* Valid credentials stored
* A refresh token

If there is no refresh token, then the credentials are cleared as normal without the revoke endpoint being called.

Otherwise, it uses its internal instance of `Authentication` to call the `revoke` endpoint, passing the refresh token. If this call is successful, the credentials are cleared. If not, the credentials remain and the callback is invoked with an error.

## Use Case

The `revoke` function could be used in most cases where the user's session is to be cleared, as it's designed to clear credentials even if there is no refresh token. Using `clear` would be beneficial if the session is to be cleared but the developer explicitly does not want to revoke the refresh token.

## Usage

The PR adds one method to the CredentialsManager API:

```swift
credentialsManager.revoke { error in
    guard error == nil else {
        print("Failed to revoke the refresh token: \(error)")
    }

    print("Token revoked and credentials cleared!")
}
```

## Testing

* [X] This change adds unit test coverage
* [X] This change has been tested on the latest version of the platform/language or why not

## Checklist

* [X] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
* [X] All existing and new tests complete without errors
* [X] All active GitHub checks have passed

### CODE REVIEW

**Comment by stevehobbsdev** [comment]
The unit tests need a little re-working to try and avoid the use of the now-deprecated `clear()` method.

**Comment by cocojoe** [comment]
@stevehobbsdev fix your tests, appear to be breaking for everything other than tvOS in Swift 5 😄 So they would also be failing locally then? Let me know if you want to pair on this.

File: Auth0/CredentialsManager.swift [hunk 1]
```diff
@@ -84,10 +84,37 @@ public struct CredentialsManager {
...
+    @available(*, deprecated, message: "use clearAndRevokeToken(callback) instead")
     public func clear() -> Bool {
         return self.storage.deleteEntry(forKey: storeKey)
     }
 
+    /// Revokes the refresh token and, if successful, the credentials are cleared. Otherwise,
+    /// the credentials are not cleared and an error is raised through the callback.
+    ///
+    /// - Parameter callback: callback with an error if the refresh token could not be revoked
+    public func clearAndRevokeToken(_ callback: @escaping (CredentialsManagerError?) -> Void) {
```
Comment (cocojoe) [inline Auth0/CredentialsManager.swift]: might be better to be explicit here in the naming `clearAndRevokeRefreshToken` although feels a bit wordy. I am wondering if this should simply be `revoke` which performs both actions. let's discuss internally

File: Auth0/CredentialsManager.swift [hunk 2]
```diff
@@ -84,10 +84,37 @@ public struct CredentialsManager {
     /// Clear credentials stored in keychain
     ///
     /// - Returns: if credentials were removed
+    @available(*, deprecated, message: "use clearAndRevokeToken(callback) instead")
```
Comment (cocojoe) [inline Auth0/CredentialsManager.swift]: In retrospect I'm not sure if this should be deprecated, there is nothing technically wrong with it as you don't have to use refresh tokens and can simply use it to securely store an AT, although that would be a less common use case.

File: Auth0/CredentialsManager.swift [hunk 3]
```diff
@@ -84,10 +84,37 @@ public struct CredentialsManager {
...
+                return callback(nil)
+        }
+
+        self.authentication
+            .revoke(refreshToken: refreshToken)
+            .start { result in
+                switch result {
+                case .failure(let error):
+                    callback(CredentialsManagerError.revokeFailed(error))
+                default:
```
Comment (cocojoe) [inline Auth0/CredentialsManager.swift]: by explicit vs using default, there are only two outcomes .failure or .success

File: Auth0/CredentialsManager.swift (hunk 2 above)
Comment (stevehobbsdev) [reply Auth0/CredentialsManager.swift]: Still having the `clear()` function would certainly make the test suite cleaner; I've had to refactor them out to remove the compiler warnings, but doing that requires a sprinkling of async handling code throughout the tests.

Then it's just communicating to the user best practises around whether to use `clear()` or `revoke()` depending on their situation.

File: App/Assets.xcassets/AppIcon.appiconset/Contents.json [hunk 4]
```diff
@@ -39,6 +39,11 @@
       "idiom" : "iphone",
       "size" : "60x60",
       "scale" : "3x"
+    },
+    {
+      "idiom" : "ios-marketing",
```
[the remark this answers is not in the document]
Comment (stevehobbsdev) [reply App/Assets.xcassets/AppIcon.appiconset/Contents.json]: Not sure, I think this might have been automatically added by XCode when I opened the project. Have reverted.

File: App/Base.lproj/Main.storyboard [hunk 5]
```diff
@@ -1,10 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="11542" systemVersion="16B2555" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" colorMatched="YES" initialViewController="BYZ-38-t0r">
-    <device id="retina4_7" orientation="portrait">
-        <adaptation id="fullscreen"/>
-    </device>
+<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="14868" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" colorMatched="YES" initialViewController="BYZ-38-t0r">
+    <device id="retina4_7" orientation="portrait" appearance="light"/>
     <dependencies>
-        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="11524"/>
+        <deployment identifier="iOS"/>
```
[the remark this answers is not in the document]
Comment (stevehobbsdev) [reply App/Base.lproj/Main.storyboard]: Not sure, I think this might have been automatically added by XCode when I opened the project. Have reverted.

File: Auth0/CredentialsManager.swift (hunk 3 above)
Comment (stevehobbsdev) [reply Auth0/CredentialsManager.swift]: I've removed the default case in favour of `.success`

File: Auth0/CredentialsManager.swift (hunk 1 above)
Comment (stevehobbsdev) [reply Auth0/CredentialsManager.swift]: This has now been renamed to `revoke`.

File: Auth0/CredentialsManager.swift (hunk 2 above)
Comment (stevehobbsdev) [reply Auth0/CredentialsManager.swift]: I've removed the deprecation on this

File: Auth0/CredentialsManager.swift [hunk 6]
```diff
@@ -88,6 +88,32 @@ public struct CredentialsManager {
         return self.storage.deleteEntry(forKey: storeKey)
     }
 
+    /// Revokes the refresh token and, if successful, the credentials are cleared. Otherwise,
+    /// the credentials are not cleared and an error is raised through the callback.
```
Comment (lbalmaceda) [inline Auth0/CredentialsManager.swift]: - Mention the endpoint being called (revoke token)
- Modify the doc. Explain how it will behave when the refresh token is not present. (will clear credentials and not throw errors)

File: Auth0Tests/CredentialsManagerSpec.swift [hunk 7]
```diff
@@ -73,6 +73,76 @@ class CredentialsManagerSpec: QuickSpec {
...
+                waitUntil(timeout: 2) { done in
+                    credentialsManager.revoke {
+                        expect($0).to(beNil())
+                        expect(credentialsManager.hasValid()).to(beFalse())
+                        done()
+                    }
+                }
+            }
+            
+            it("should not return an error if there is no refresh token") {
```
Comment (lbalmaceda) [inline Auth0Tests/CredentialsManagerSpec.swift]: "but clear credentials anyway"

File: Auth0/CredentialsManager.swift (hunk 6 above)
Comment (stevehobbsdev) [reply Auth0/CredentialsManager.swift]: @lbalmaceda I've updated this but it's not showing on the diff for some reason:

https://github.com/auth0/Auth0.swift/blob/creds-manager-revoke-token/Auth0/CredentialsManager.swift#L91
